### PR TITLE
Dice genetic optimizations

### PR DESCRIPTION
**Pull request by soundarya98**
- Fixed bug: posthoc sparsity throws errors when final_cfs_df is None
- Fixed bug: desired_class = "opposite" wasn't getting initialized properly
- Added check that generated CFs are from the desired_class / desired_range
- Refactored genetic algorithm - vectorized losses, predictions, switched to numpy arrays for efficiency
- Removed diversity loss and changed code from O(population_size * total_CFs) to O(population_size)
- Added tests: Checking that we return valid cfs when the algorithm doesn't converge

### CODE REVIEW

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 1]
```diff
@@ -94,94 +94,75 @@ def do_loss_initializations(self, yloss_type, diversity_loss_type, feature_weigh
...
+            if kx >= len(cfs):
+                break
+            one_init = np.zeros(self.data_interface.number_of_features)
+            for jx, feature in enumerate(self.data_interface.feature_names):
+                if feature not in features_to_vary:
+                    one_init[jx] = (query_instance[jx])
+                else:
+                    if feature in self.data_interface.continuous_feature_names:
+                        if self.feature_range[feature][0] <= cfs.iloc[kx][jx] <= self.feature_range[feature][1]:
+                            one_init[jx] = cfs.iloc[kx][jx]
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:140]: use `at` or `iat` instead of loc/iloc everywhere in the code. It is much faster. https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html _"For production code, we recommended that you take advantage of the optimized pandas data access methods exposed in this chapter."_

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 2]
```diff
@@ -94,94 +94,75 @@ def do_loss_initializations(self, yloss_type, diversity_loss_type, feature_weigh
...
-                                if self.feature_range[feature][0] <= query_instance[0][jx] <= self.feature_range[feature][1]:
-                                    one_init[0].append(query_instance[0][jx])
-                                else:
-                                    one_init[0].append(
-                                        np.random.uniform(self.feature_range[feature][0],
-                                                          self.feature_range[feature][1]))
+                                one_init[jx] = np.random.uniform(self.feature_range[feature][0], self.feature_range[feature][1])
+                    else:
+                        if cfs.iloc[kx][jx] in self.feature_range[feature]:
+                            one_init[jx] = cfs.iloc[kx][jx]
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:148]: same comment for iloc. search and replace everywhere.

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 3]
```diff
@@ -227,11 +208,11 @@ def do_param_initializations(self, total_CFs, initialization, desired_range, des
         self.do_loss_initializations(yloss_type, diversity_loss_type, feature_weights, encoding='label')
         self.update_hyperparameters(proximity_weight, diversity_weight, categorical_penalty)
 
-    def _generate_counterfactuals(self, query_instance, total_CFs, initialization="kdtree", desired_range=None, desired_class="opposite", proximity_weight=0.5,
+    def _generate_counterfactuals(self, query_instance, total_CFs, initialization="kdtree", desired_range=None, desired_class="opposite", proximity_weight=0.1,
                                  diversity_weight=5.0, categorical_penalty=0.1, algorithm="DiverseCF",
                                  features_to_vary="all", permitted_range=None, yloss_type="hinge_loss",
                                  diversity_loss_type="dpp_style:inverse_dist", feature_weights="inverse_mad", stopping_threshold=0.5, posthoc_sparsity_param=0.1, posthoc_sparsity_algorithm="binary",
-                                 maxiterations=10000, verbose=False):
+                                 maxiterations=500, thresh=1e-2, verbose=False):
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:215]: descriptive names are useful. how about 'stopping_thresh' or 'convergence_thresh'

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 4]
```diff
@@ -305,151 +288,78 @@ def predict_fn_scores(self, input_instance):
...
-                    yloss += temp_loss
-            return yloss / self.total_CFs
+                yloss = np.zeros(len(predicted_value))
+                for i in range(len(predicted_value)):
+                    if not desired_range[0] <= predicted_value[i] <= desired_range[1]:
+                        yloss[i] = min(abs(predicted_value[i] - desired_range[0]), abs(predicted_value[i] - desired_range[1]))
+            return yloss
 
-    def compute_dist(self, x_hat, x1):
+    def compute_proximity_loss(self, x_hat):
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:316]: what about cont versus categorical features? does the same loss work?
if so, good to write a comment in line 316 on how it works for both kinds of features.

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 5]
```diff
@@ -305,151 +288,78 @@ def predict_fn_scores(self, input_instance):
...
+        one_init = np.zeros(self.data_interface.number_of_features)
+        for j in range(self.data_interface.number_of_features):
+            gp1 = k1[j]
+            gp2 = k2[j]
+            feat_name = self.data_interface.feature_names[j]
 
-            child_chromosome.append(np.array(one_init))
-        return child_chromosome
+            # random probability
+            prob = random.random()
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:340]: for consistency, can use np.random and avoid importing python's default random. They have different seeds.

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 6]
```diff
@@ -461,47 +371,64 @@ def find_counterfactuals(self, query_instance, desired_range, desired_class, fea
...
-            cfs_preds = [self.predict_fn(cfs) for cfs in current_best_cf]
-            # 10% of the next generation is fittest members of current generation
-            population_fitness = sorted(population_fitness, key=lambda x: x[1])
-            s = int((10 * self.population_size) / 100)
-            new_generation = [population[tup[0]] for tup in population_fitness[:s]]
-
-            # 90% of the next generation obtained from top 50% of fittest members of current generation
-            s = int((90 * self.population_size) / 100)
+
+            new_array = [tuple(row) for row in population]
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:388]: better not to mix python lists and np arrays. Since you convert to np array in the next line, can define `new_array` as a np.array too.

Another faster, pythonic option is to simply do `new_array=set(population)` and you will automatically get unique values. Then convert it to back to list or np.array, as `list(new_array)`

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 7]
```diff
@@ -94,94 +94,75 @@ def do_loss_initializations(self, yloss_type, diversity_loss_type, feature_weigh
...
+                    if feature in self.data_interface.continuous_feature_names:
+                        one_init[jx] = np.round(np.random.uniform(self.feature_range[feature][0], self.feature_range[feature][1]), precisions[jx])
                     else:
-                        temp_cfs.append(np.array(one_init))
-                ix += 1
-            self.cfs.append(temp_cfs)
+                        one_init[jx] = np.random.choice(self.feature_range[feature])
+                else:
+                    one_init[jx] = query_instance[jx]
+            if self.model.model_type == 'classifier':
```
[the remark this answers is not in the document]
Comment (amit-sharma) [reply dice_ml/explainer_interfaces/dice_genetic.py:111]: also can use decide_cf_validity from explainer_base to check whether cf is valid. avoid repetitive code.

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 8]
```diff
@@ -461,47 +371,64 @@ def find_counterfactuals(self, query_instance, desired_range, desired_class, fea
...
-        current_best_cf = []
         stop_cnt = 0
         cfs_preds = [np.inf]*self.total_CFs
-        while iterations < maxiterations:
-            if abs(previous_best_loss - current_best_loss) <= 1e-2: #and (self.model.model_type == 'classifier' and all(i == desired_class for i in cfs_preds) or (self.model.model_type == 'regressor' and all(desired_range[0] <= i <= desired_range[1] for i in cfs_preds))):
+        to_pred = None
+
+        while iterations < maxiterations or len(population) == self.total_CFs:
+            start = timeit.default_timer()
+            if abs(previous_best_loss - current_best_loss) <= thresh and (self.model.model_type == 'classifier' and all(i == desired_class for i in cfs_preds) or (self.model.model_type == 'regressor' and all(desired_range[0] <= i <= desired_range[1] for i in cfs_preds))):
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:380]: the second two conditions can be checked by simply invoking `decide_cf_validity` from explainer_base.

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 9]
```diff
@@ -461,47 +371,64 @@ def find_counterfactuals(self, query_instance, desired_range, desired_class, fea
...
+            population = np.concatenate([new_generation_1, new_generation_2])
             iterations += 1
 
-        self.final_cfs = current_best_cf
-        self.cfs_preds = [self.predict_fn(cfs) for cfs in self.final_cfs]
+        self.cfs_preds = []
+        self.final_cfs = []
+        if self.model.model_type == 'classifier':
+            i = 0
+            while i < self.total_CFs:
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:418]: the next few lines can be simplified by using decide_cf_validity from explainer_base.

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 10]
```diff
@@ -461,47 +371,64 @@ def find_counterfactuals(self, query_instance, desired_range, desired_class, fea
...
+            s = self.population_size - s
+            new_generation_2 = np.zeros((s, self.data_interface.number_of_features))
             for _ in range(s):
-                parent1 = random.choice(population[:int(50 * self.population_size / 100)])
-                parent2 = random.choice(population[:int(50 * self.population_size / 100)])
+                parent1 = random.choice(population[:int(len(population) / 2)])
+                parent2 = random.choice(population[:int(len(population) / 2)])
                 child = self.mate(parent1, parent2, features_to_vary, query_instance)
-                new_generation.append(child)
+                new_generation_2[_] = child
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:409]: usually `_` is only used if you don't want to use the variable.  here use something like i or j

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 11]
```diff
@@ -517,18 +444,25 @@ def find_counterfactuals(self, query_instance, desired_range, desired_class, fea
         else:
             self.final_cfs_df_sparse = None
 
-        # to display the values with the same precision as the original data
-        precisions = self.data_interface.get_decimal_precisions()
-        for ix, feature in enumerate(self.data_interface.continuous_feature_names):
-            self.final_cfs_df[feature] = self.final_cfs_df[feature].astype(float).round(precisions[ix])
-            self.final_cfs_df_sparse[feature] = self.final_cfs_df_sparse[feature].astype(float).round(precisions[ix])
+        if self.final_cfs_df is not None:
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:447]: precision-based output should be a part of explainer_base and common to all methods.

**Review by amit-sharma — COMMENTED** [review]
Great to see the optimizations. Inline comments added. A major one is to avoid check cf-class validity multiple times in the code and use a function from explainer-base

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 12]
```diff
@@ -96,7 +96,7 @@ def _generate_counterfactuals(self, query_instance, total_CFs, desired_range=Non
         self.dataset_with_predictions, self.KD_tree, self.predictions = self.build_KD_tree(data_df_copy, desired_range,
                                                                                            desired_class, self.predicted_outcome_name)
 
-        query_instance, final_cfs, cfs_preds = self.find_counterfactuals(data_df_copy,
+        query_instance, cfs_preds = self.find_counterfactuals(data_df_copy,
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:99]: The `find_counterfactuals` needs to be refactored. This is a larger problem with other files also--perhaps we can take this up in a future PR. Ideally it should return the final_cfs and then we should assign it to self.final_cfs here in this function. The find_counterfactuals shouldn't be setting `self` attributes.

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 13]
```diff
@@ -402,33 +379,27 @@ def find_counterfactuals(self, query_instance, desired_range, desired_class, fea
             # rest of the next generation obtained from top 50% of fittest members of current generation
             s = self.population_size - s
             new_generation_2 = np.zeros((s, self.data_interface.number_of_features))
-            for _ in range(s):
+            for i in range(s):
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:382]: I know I had suggested `i` here,but looking at a few lines below at 394 is also using `i`, it feels a bit dangerous. You can perhaps use a different variable name here to avoid confusion.

File: dice_ml/explainer_interfaces/dice_random.py [hunk 14]
```diff
@@ -85,7 +85,7 @@ def _generate_counterfactuals(self, query_instance, total_CFs, desired_range,  d
             self.fixed_features_values = {}
             for feature in self.data_interface.feature_names:
                 if feature not in features_to_vary:
-                    self.fixed_features_values[feature] = query_instance[feature].iloc[0]
+                    self.fixed_features_values[feature] = query_instance[feature].iat[0]
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_random.py:88]: thanks for changing this in other files too!

File: dice_ml/explainer_interfaces/explainer_base.py [hunk 15]
```diff
@@ -575,3 +591,11 @@ def build_KD_tree(self, data_df_copy, desired_range, desired_class, predicted_ou
             KD_tree = KDTree(dummies)
 
         return dataset_with_predictions, KD_tree, predictions
+
+    def round_to_precision(self):
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/explainer_base.py:595]: nice, thanks for adding this common function

**Review by amit-sharma — COMMENTED** [review]
thanks for the quick turnaround @soundarya98 I added just a few minor comments. There is also some tests failing. Once those are addressed, I'm happy to merge

File: dice_ml/explainer_interfaces/dice_KD.py (hunk 12 above)
Comment (soundarya98) [reply dice_ml/explainer_interfaces/dice_KD.py:99]: Yes, I agree completely! Functions ideally shouldn't change any self variables and should return everything that is changed
